src/Bicep.Core/Configuration/ConfigurationManager.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Reflection;
+using System.Security;
+using System.Text.Json;
+
+namespace Bicep.Core.Configuration
+{
+    public class ConfigurationManager : IConfigurationManager
+    {
+        public const string BuiltInConfigurationResourceName = "Bicep.Core.Configuration.bicepconfig.json";
+
+        private static readonly IConfiguration builtInRawConfiguration = BuildBuiltInRawConfiguration();
+
+        private static readonly Lazy<RootConfiguration> builtInConfigurationLazy =
+            new(() => RootConfiguration.Bind(builtInRawConfiguration, BuiltInConfigurationResourceName));
+
+        private static readonly Lazy<RootConfiguration> builtInConfigurationWithAnalyzersDisabledLazy =
+            new(() => RootConfiguration.Bind(builtInRawConfiguration, BuiltInConfigurationResourceName, disableAnalyzers: true));
+
+        private readonly IFileSystem fileSystem;
+
+        public ConfigurationManager(IFileSystem fileSystem)
+        {
+            this.fileSystem = fileSystem;
+        }
+
+        public RootConfiguration GetBuiltInConfiguration(bool disableAnalyzers = false) => disableAnalyzers
+            ? builtInConfigurationWithAnalyzersDisabledLazy.Value
+            : builtInConfigurationLazy.Value;
+
+        public RootConfiguration GetConfiguration(Uri sourceFileUri)
+        {
+            var customConfigurationPath = DiscoverConfigurationFile(fileSystem.Path.GetDirectoryName(sourceFileUri.LocalPath));
+
+            if (customConfigurationPath is not null)
+            {
+                IConfiguration? customConfiguration;
+
+                try
+                {
+                    using var stream = fileSystem.FileStream.Create(customConfigurationPath, FileMode.Open, FileAccess.Read);
+                    var builder = new ConfigurationBuilder()
+                        .AddInMemoryCollection(builtInRawConfiguration.AsEnumerable())
+                        .AddJsonStream(stream);
```

Comment:
>AddJsonStream [](http://example.com/codeflow?start=25&length=13)

I'm assuming this does a merge of the file with the default config. 

What are the semantics on merging objects? Do objects get replaced or does it combine the properties? And is it recursive? Do property values get replaced or merged as well?

Reply:
The semantic is patch on merging objects. It combines properties and it is recursive. Any property value will get replaced if the value type is `string`, `int`, `bool`, or `null`. For `array`, the behavior is a bit weird - each array item in the original config will get replaced by the new array item in the new config in order. E.g.: 

```
// old config
{
  "numbers": [1, 2, 3, 4]
}

// new config
{
  "numbers": [0, 0]
}

// merged config
{
  "numbers": [0, 0, 3, 4]
}
```

Reply:
Yeah I think you're right. Is it possible to customize that in the config library?